cpp/include/raft_runtime/neighbors/ivf_flat.hpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/neighbors/ivf_flat_types.hpp>
+
+namespace raft::runtime::neighbors::ivf_flat {
+
+// We define overloads for build and extend with void return type. This is used in the Cython
+// wrappers, where exception handling is not compatible with return type that has nontrivial
+// constructor.
+#define RAFT_INST_BUILD_EXTEND(T, IdxT)                                      \
+  auto build(raft::device_resources const& handle,                           \
+             raft::device_matrix_view<const T, uint64_t, row_major> dataset, \
+             const raft::neighbors::ivf_flat::index_params& params)          \
+    ->raft::neighbors::ivf_flat::index<T, IdxT>;                             \
+                                                                             \
+  auto extend(raft::device_resources const& handle,                          \
+              const raft::neighbors::ivf_flat::index<T, IdxT>& orig_index,   \
+              const T* new_vectors,                                          \
+              const IdxT* new_indices,                                       \
+              IdxT n_rows)                                                   \
+    ->raft::neighbors::ivf_flat::index<T, IdxT>;                             \
+                                                                             \
+  void build(raft::device_resources const& handle,                           \
+             raft::device_matrix_view<const T, uint64_t, row_major> dataset, \
+             const raft::neighbors::ivf_flat::index_params& params,          \
+             raft::neighbors::ivf_flat::index<T, IdxT>* idx);                \
+                                                                             \
+  void extend(raft::device_resources const& handle,                          \
+              raft::neighbors::ivf_flat::index<T, IdxT>* idx,                \
```

Comment:
Currently I use the pointer based API of extend. If std::optional is compatible with Cython, then we could switch to the mdspan API of extend https://github.com/rapidsai/raft/blob/e7f0268c9f5c17756c415ed47e1856030b3f235d/cpp/include/raft/neighbors/ivf_flat.cuh#L291-L294